### PR TITLE
Add exporter tests

### PR DESCRIPTION
**Pull request by atodorov**
Starting with a metadata.db and cs.repo with several known packages
we try the following scenarios

- export a non-existing RPM package: should give an error
- export an existing RPM pacakge: content should be the same
  as if we tried an ostree checkout of the commit which imported
  that very same package. To make it easier on us this is the last
  imported package where we just `checkout master`.
- export two existing RPM packages which are in conflict with one
  another. Still not sure what needs to happen here.

### CODE REVIEW

**Comment by atodorov** [comment]
NOTE: the test will fail at line 50 b/c the exporter always returns a zero exit code even if it errors out because it can't find some groups.

NOTE2: I'm having troubles with the test for conflicting RPMs. For some reason it can't find the tog-pegasus-libs group:

```
$ sudo ./dist/build/export/export ./export_metadata.db ./export.repo/ ./exported-content.d/ filesystem-3.2-21.el7.x86_64 tog-pegasus-libs-2.14.1-3.el7.x86_64 libcmpiCppImpl0-2.0.3-5.el7.x86_64
"No such group tog-pegasus-libs-2.14.1-3.el7.x86_64"
```

**Comment by atodorov** [comment]
[TESTS for exporter: manifest to output directory](https://trello.com/c/mDD0gnDp/358-tests-for-exporter-manifest-to-output-directory)

**Comment by clumens** [comment]
So far, so good on the tests.

I would be happy if you wanted to add a patch to this set to do something about the exit code problem.  If you need pointers on that, just ask.

As for conflicting RPMs, that's a good question.  I don't have any idea what should happen there either.  We should have conflicts info in the database, so we can detect them there.  But what about when you export conflicting RPMs in different operations?  Open to suggestions.

Finally, on tog-pegasus-libs... my first thought is that it's not a problem with the database but with something else that is getting handled by that same error message.  That is, tog-pegasus-libs-whatever is in the database but that heinous query in nevraToGroupId is wrong somehow (I sure hope not) or we're not splitting the NEVRA up right so we are querying for something that doesn't exist.  Testing splitFilename with that package name would be a good first step.

**Comment by atodorov** [comment]
> I would be happy if you wanted to add a patch to this set to do something about the exit code problem. If you need pointers on that, just ask.

```
getGroupIdC :: (MonadError String m, MonadBaseControl IO m, MonadIO m) => Conduit T.Text (SqlPersistT m) (Key Groups)
getGroupIdC = await >>= \case
    Nothing    -> return ()
    Just thing -> do
        maybeId <- lift $ nevraToGroupId (splitFilename thing)
        case maybeId of
            Just gid -> yield gid >> getGroupIdC
            Nothing  -> throwError $ "No such group " ++ T.unpack thing
```

So I see this error comes from the last line in this function but I have no idea how to set the exit code in this case. Can you give me a hint?

**Comment by atodorov** [comment]
> Testing splitFilename with that package name would be a good first step.

That works fine but this package has an epoch which also needs to be specified on the command line for `export`. Still not sure which function is that (will check and add test if I can) but it is not related to the current PR.

I've added the tests for conflicting packages and I'm happy to say that exporter gives an error (file exists) which is great.

**Comment by atodorov** [comment]
@clumens rebased to latest master. All tests now pass. How do you feel about the exit code change ?

File: importer/tools/export.hs [hunk 1]
```diff
@@ -84,8 +89,8 @@ main = do
     let things = map T.pack $ match !! 0 : otherThings
 
     let (handler, objectSink) = if ".tar" `isSuffixOf` out_path
-            then (\e -> print e >> whenM (doesFileExist out_path) (removeFile out_path), CS.objectToTarEntry .| Tar.tarSink out_path)
-            else (print, Directory.directorySink out_path)
+            then (\e -> printAndDie e >> whenM (doesFileExist out_path) (removeFile out_path), CS.objectToTarEntry .| Tar.tarSink out_path)
+            else (printAndDie, Directory.directorySink out_path)
```
Comment (dashea) [inline importer/tools/export.hs]: Running printAndDie before the other actions (the tar file cleanup) means that the program will exit before the cleanup runs. Maybe move the "andDie" part to the error handler. The last line of main could be something like `whenLeft result (handler >> exitFailure)`

**Comment by atodorov** [comment]
hmm, now test failed with:

export: ./exported-content.d/etc/shadow: openFile: does not exist (No such file or directory)

This is something new, haven't seen it before. Any idea why ?

**Comment by atodorov** [comment]
ok, I've fixed problems related to setup and systemd-sysusers command. Now the tests are failing at the point where I try to export two conflicting packages. There doesn't seem to be any output and the return code is 0, hence the failure. A few days ago the exporter did spew an error about not being able to create symbolic links.

**Comment by atodorov** [comment]
ok, I've fixed problems related to setup and systemd-sysusers command. Now the tests are failing at the point where I try to export two conflicting packages. There doesn't seem to be any output and the return code is 0, hence the failure. A few days ago the exporter did spew an error about not being able to create symbolic links.

**Pull request merged**